### DISCUSSION TITLE
Error compiling source after issuing 'ASCII'

### DISCUSSION THREAD

**Discussion by ghost** [kej715/DtCyber#127] [Q&A] [answered]
Hi all,

please correct me if I'm wrong, but to be able to write lower case to a file I need to use the command _ASCII_, right? At least I see that otherwise my input is automatically converted to upper case.

When writing code, for example in Pascal, I have problems to compile code that has been written after _ASCII_. Specifically I get an _UNEXPECTED SYMBOL_ for the colon symbols (:). Same code written without  _ASCII_, compiles without any issues. But I cannot have for example messages displayed in lower case.

Is this a bug? Am I doing something wrong?

Thanks a lot in advance.

**Comment by kej715** [comment, marked as answer]
Hi Dave,

The ASCII and NORMAL commands affect how IAF interacts with NAM and 1TM. It
also affects the operation of a few other tools such as text edits, but It
does not affect the operation of compilers.

NOS has multiple character set encodings. The default and most frequently
used one is called "Display Code". Display Code is encoded as 6 bits per
character and supports upper case only. Another encoding is called
"Extended Display Code" (aka 6/12 Extended Display Code). Extended Display
Code encodes characters in 6 bits or 12 bits. It reserves two 6-bit
characters as escape characters. The two 6-bit escape characters alter the
interpretation of the immediately following 6-bit character. Thus, an
"ordinary" character is 6-bits, and an escaped one is 12. Extended Display
Code supports lower case and has mappings for the full 128-bit ASCII
repertoire. It has some anomalies too, e.g., it has two representations for
":" (colon).

The two escape characters in the Extended Display Code encoding are "^" and
"@". Lower case letters are indicated by "^" (e.g., "^A"), and "^" is also
used in encoding ASCII control characters as well as a few other characters
not supported by Display Code. The escape characters themselves (and a
small handful of others) are encoded using "@" as the escape character
(e.g., "@" itself is encoded as ***@***.***", and ":" is usually encoded as ***@***.***").
The Display Code and Extended Display Code encodings, along with their
mappings onto ASCII, are documented in Appendix A of the "NOS 2 Volume 3
System Commands" manual.

Other encodings supported by NOS include one where "true" ASCII characters
are encoded right-justified in 12 bits and packed 5 per 60-bit word, and
another similar encoding represents "true" ASCII characters in 8 bits
packed 7.5 per word (i.e., 15 ASCII characters packed into two words where
one character spans the two words).

The ASCII command conditions IAF to process text using the Extended Display
Code encoding. IAF always exchanges "true" ASCII with NAM. The ASCII
command conditions It to apply an Extended Display Code mapping when
exchanging text with NAM, and the NORMAL command (the default mode)
conditions it to apply a Display Code mapping. The Extended Display Code
mapping produces mixed case and the full ASCII repertoire. The Display Code
mapping is the usual default, however, and it produces upper case only and
a 64-character subset of the repertoire.

The text editors FSE and XEDIT recognize the character encoding mode set by
the ASCII and NORMAL commands, and they will recognize commands entered in
lower case (i.e., internally encoded using the Extended Display Code escape
characters) when ASCII mode is in effect. Almost -NONE- of the compilers
recognize ASCII mode. Almost all of the compilers expect ONLY ordinary
Display Code. Of course, they -will- accept quoted strings containing the
Extended Display Code escape characters because the escape characters are
members of the ordinary Display Code character set, so you can write
programs that output mixed text, and you can also write programs that
accept mixed text provided that your programs handle the escape codes
explicitly.

The one current exception is that the CYBIL compiler expects ONLY Extended
Display Code.

cheers,
Kevin

**Reply by ghost** [reply]
Thanks a lot @kej715 for the greatly detailed explanation. I now understand better how this works. And also now I understand why there were all those ^ symbols in the INDEX file of the user WWW. I thought it was something only related with the HTF server, so I didn't connect the dots.

**Comment by KC5VDJ** [comment]
I think the Pascal,lfn,opts/A+ could possibly help if it is a detection
issue, like "as" with xedit.

**Comment by kej715** [comment]
Nope. The A+ / A- option selects between the ASCII general and ASCII
scientific character sets and applies when the compiler has been built to
accept Extended Display Code. The compiler included in the NOS. 2.8.7
ready-to-run packages was -not- built to accept Extended Display Code.

cheers,
Kevin

**Comment by KC5VDJ** [comment]
Yup.  My bad.  Pascal-6000 requires 7/12 ASCII iirc.  Big diff.  Can't find it at the moment, but iirc it is in library/pasdoc.
